libnmstate/nm/connection.py

```diff
@@ -146,15 +158,21 @@ def safe_activate_async(self):
         cancellable = self._mainloop.new_cancellable()
 
         active_conn = get_device_active_connection(self.nmdevice)
+        nm_acsreason = nmclient.NM.ActiveConnectionStateReason
         if active_conn:
             ac = ActiveConnection(active_conn)
-            if ac.is_activating:
+            if (
+                ac.is_activating
+                or ac.reason == nm_acsreason.DEVICE_DISCONNECTED
+            ):
                 logging.debug(
-                    'Connection activation in progress: dev=%s, state=%s',
+                    'Connection activation in progress: dev=%s, state=%s, '
+                    'reason=%s',
                     ac.devname,
                     ac.state,
+                    ac.reason,
                 )
-                self.waitfor_active_connection_async(ac)
+                self.wait_active_async()
```

Comment:
I think you need to "sleep" only in case it is "disconnected". In case it is "activating", we just need to wait for the state to change to activated.

Reply:
The `wait_active_async()` just register the `_active_timeout_callbak()` to be called repeatedly until it return True.
I don't use the old way(waiting activate connection state change) as the active connection can became None sometimes.